### DISCUSSION TITLE
How to write multi-line condition in if

### DISCUSSION THREAD

**Discussion by kei-yamazaki** [community/community#25641] [Actions]
<p>I used <code>'&gt;'</code> to describe a multi-line condition in if.</p>
<p>I thought this would work as expected, but the combination of <code>'&gt;'</code> and <code>'${{ }}'</code> seems to work strangely.</p>
<p>Below is the workflow created for confirmation.</p>
<pre><code class="lang-auto">name: condition-tests

on:
  push:
    branches: [ master ]

jobs:
  test:
    runs-on: ubuntu-latest

    steps:
    - name: test1
      run: echo hoge
      if: ( 'a' == 'a' )

    - name: test2
      run: echo hoge
      if: ( 'a' == 'b' )

    - name: test3
      run: echo hoge
      if: &gt;
        'a' == 'a'

    - name: test4
      run: echo hoge
      if: &gt;
        'a' == 'b'

    - name: test5
      run: echo hoge
      if: &gt;
        ${{ 'a' == 'a' }}

    - name: test6
      run: echo hoge
      if: &gt;
        ${{ 'a' == 'b' }}
</code></pre>
<p>And the execution result is this.<br>
</p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/2/24f59e0a8b8b10befcbfd724a1c4e9900ea2200b.png" data-download-href="/uploads/short-url/5gXq8VIfqDpZyB4MV4zEW2uO5Tl.png?dl=1" title="image" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/1715090/181092319-78f715f8-c40f-4034-b220-c58bbca78a51.png" alt="image" data-base62-sha1="5gXq8VIfqDpZyB4MV4zEW2uO5Tl" width="295" height="499" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/2/24f59e0a8b8b10befcbfd724a1c4e9900ea2200b_2_295x499.png, https://aws1.discourse-cdn.com/github/original/2X/2/24f59e0a8b8b10befcbfd724a1c4e9900ea2200b.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/2/24f59e0a8b8b10befcbfd724a1c4e9900ea2200b.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/2/24f59e0a8b8b10befcbfd724a1c4e9900ea2200b_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">image</span><span class="informations">434×734 20.6 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>I thought that the last test6 was skipped, but it is not actually skipped.</p>
<p>I think that there is something wrong with the Parser in GitHub Actions when I combine <code>'&gt;'</code> and <code>'${{ }}'</code>. Is my workflow something syntactically wrong ?</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/kei-yamazaki">@kei-yamazaki</a>,</p>
<p>you need to use the symbol ‘<strong>|</strong>’ to set the multi-line condition.<br>
For example:</p>
<pre><code class="lang-yaml">jobs:
  job1:
    if: |
      github.event_name == 'push' ||
      (github.event_name == 'pull_request' &amp;&amp; github.head_ref == 'master')
</code></pre>
<p>Here is a same topic as reference:</p><aside class="quote quote-modified" data-post="2" data-topic="116903">
  <div class="title">
    <div class="quote-controls"></div>
    
    <a href="https://github.community/t/github-action-combining-if-condition-or-writing-it-in-multiple-lines/116903/2">Github action combining if condition or writing it in Multiple lines</a> <a class="badge-wrapper  bullet" href="/c/code-to-cloud/github-actions/41"><span class="badge-category-parent-bg" style="background-color: #6a737d;"></span><span class="badge-category-bg" style="background-color: #2188ff;"></span><span style="" data-drop-close="true" class="badge-category clear-badge" title="Discussion and questions about GitHub Actions—automation for all of your development workflows, now with world-class CI/CD.">GitHub Actions</span></a>
  </div>
  <blockquote>
    <a class="mention" href="/uditgaurav">@uditgaurav</a>, 
You can write like as the following: 
- name: Running all tests
  if: |
    startsWith(github.event.comment.body, '/run-e2e-all') ||
    startsWith(github.event.comment.body, '/run-e2e-first-test') ||
    startsWith(github.event.comment.body, '/run-e2e-second-test') ||
    startsWith(github.event.comment.body, '/run-e2e-third-test') ||
    startsWith(github.event.comment.body, '/run-e2e-fourth-test') ||
    startsWith(github.event.comment.body, '/run-e2e-fifth-test') ||
    startsW…
  </blockquote>
</aside>

<p><strong>Note:</strong> if you use “<strong>${{ }}</strong>” to surround the condition, this will not work.</p>

**Reply by andyfeller** [reply]
The use of pipes `|` for multi-line YAML values is covered in [YAML specification for block style productions](https://yaml.org/spec/1.2.2/#chapter-8-block-style-productions).

**Comment by kei-yamazaki** [comment]
<p><a class="mention" href="/brightran">@brightran</a>,<br>
Thank you for your reply.</p>
<p>Is it a specification that doesn’t work when using <strong>“|”</strong> and <strong>“${{ }}”</strong> together?<br>
If so, I think it should be documented.<br>
<a href="https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#about-contexts-and-expressions" class="onebox" target="_blank" rel="nofollow noopener">https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#about-contexts-and-expressions</a></p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/kei-yamazaki">@kei-yamazaki</a> ,</p>
<p>Yes, I agree.<br>
The documentation even does not mention using the ‘<strong>|</strong>’ symbol to declare a multi-line <strong>if</strong> conditional.</p>

**Comment by smac89** [comment]
<p>Just an FYI for anyone wondering:</p>
<p>Multiline condition works for assignment. So something like</p>
<pre><code class="lang-yaml">jobs:
  job1:
    outputs:
      my-output: ${{ steps.get-latest-release.outcome == 'success' &amp;&amp;
                     steps.get-latest-dev.outcome == 'success' &amp;&amp;
                     (steps.get-latest-release.outputs.result != steps.get-latest-dev.outputs.result) }}
</code></pre>
<p>HTH</p>

**Comment by frewilhelm** [comment]
<p><em>[FYI to people that just found this  discussion]</em></p>
<blockquote>
<p><strong>Note:</strong> if you use “ <strong>${{ }}</strong> ” to surround the condition, this will not work.</p>
</blockquote>
<p>In the meantime it seems to work to combine <code>|</code> and <code>${{ ... }}</code> in multiline if-conditions (<em>on GHE 3.0.10</em>):</p>
<p>Example:</p>
<pre><code class="lang-auto">jobs:
  job_test:
    if: |
        ${{ github.event_name == 'push'
            || ( github.event_name == 'workflow_run'
                 &amp;&amp; github.event.workflow_run.conclusion == 'success'
               ) }}
</code></pre>

**Reply by chrisba11** [reply]
I've found that this does not actually work for me. Here are my findings:

This works as expected:
```
    if: |
      !cancelled() && needs.deploy.result == 'success' && 
      needs.tf-apply.result != 'failure'
```

This also works as expected:
```
    if: ${{ !cancelled() && needs.deploy.result == 'success' && needs.tf-apply.result != 'failure' }}
```

This does not get skipped when the workflow is cancelled and will run even when the result of the deploy job is not 'success':
```
    if: |
      ${{ !cancelled() && needs.deploy.result == 'success' && 
      needs.tf-apply.result != 'failure' }}
```

**Comment by existere** [comment]
<aside class="quote no-group quote-modified" data-username="fwilhelm92" data-post="7" data-topic="128477" data-full="true">
<div class="title">
<div class="quote-controls"></div>
 fwilhelm92:</div>
<blockquote>
<p><em>[FYI to people that just found this  discussion]</em><br>
In the meantime it seems to work to combine <code>|</code> and <code>${{ ... }}</code> in multiline if-conditions (<em>on GHE 3.0.10</em>):</p>
</blockquote>
</aside>
<p>This did not work for me. In my case the first conditional was evaluated and did not throw an error but the second never was. In your case, having || would not throw an error either unless you tested <code>worflow_run</code> manually.</p>
<pre><code class="lang-auto">      - name: Set AWS Env
        if: |
          ${{ github.ref == 'refs/heads/master' 
            &amp;&amp; (!contains(github.event.head_commit.message, 'skip-release')
              )}}
</code></pre>
<p>However, <a class="mention" href="/brightran">@brightran</a> 's example did work:</p>
<pre><code class="lang-auto">      - name: Set AWS Env
        if: |
          github.ref != 'refs/heads/master' 
            &amp;&amp; !contains(github.event.head_commit.message, 'skip-release')
</code></pre>

**Reply by byam** [reply]
thanks, it works for me ;)

**Comment by brianjmurrell** [comment]
My testing just now indicates that only `>-` actually works when you use `${{ }}`.  Both of `>` and `|` do not work.  Here is my `if:` conditional:

```
    if: >-
        ${{ (inputs.BRANCH != '' ||
             needs.Import-commit-pragmas.outputs.run-gha ||
             inputs.RUN_GHA ||
             github.event_name == 'workflow_dispatch') &&
            needs.Import-commit-pragmas.outputs.rpm-test-version == '' &&
            !contains(needs.Import-commit-pragmas.outputs.pr-repos, inputs.NAME) }}
```

when the evaluation results in `false`, the job is only skipped if i use the `>-` syntax and is not skipped when using `>` or `|`.

Note that if I remove the surrounding `${{ }}` `|` can be used successfully.

Which I find infuriatingly stupid if I am being honest.  I've gnashed for 2 days now getting to the bottom of this.  Trying alternate multi-line syntax was not something I had even considered could be a cause of why my `if:` was not working.

**Reply by Hellseher** [reply]
It looks like GitHub Action "partly supports YAML" standard https://yaml.org/spec/1.2.2/#65-line-folding

It's vague specified in GHA documentation which version of YAML is implemented https://docs.github.com/en/actions/sharing-automations/creating-actions/metadata-syntax-for-github-actions

**Comment by kumar1708** [comment]
This is not working for me:

```
job:
 job-name:
    needs: [terraform]
    if: ${{ needs.terraform.outputs.has_changes == 2 && github.ref == 'refs/heads/master' && inputs.TF_WORKING_DIR == '*test' }}
```

After adding this condition (inputs.TF_WORKING_DIR == '*test') , workflow is not working as expected.

I have two directories: 
- ./gh-snow-test
- ./git-snow-test

So, I was trying to use `*test` to make this work, can someone guide

**Reply by BenCoden** [reply]
I am having the same issue

**Reply by Yang-33** [reply]
You can use `endsWith(inputs.TF_WORKING_DIR, 'test')` https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/evaluate-expressions-in-workflows-and-actions#endswith
